src/core/util/options.js

```diff
@@ -172,7 +172,7 @@ strats.watch = function (parentVal: ?Object, childVal: ?Object): ?Object {
     }
     ret[key] = parent
       ? parent.concat(child)
-      : [child]
+      : Array.isArray(child) ? child : [child]
```

Comment:
So I finally took a proper look at this issue - since the error is caused by native `watch` on `Object.prototype`  (which is a function), I think the proper fix would be adding the following at the top of the function:

``` js
if (typeof parentVal === 'function') {
  parentVal = null
}
if (typeof childVal === 'function') {
  childVal = null
}
```

Essentially, ignore if the value is `Object.prototype.watch`. The test case will need to be updated accordingly.